vision/api/VisionTest/ProductSearchTest.cs

```diff
@@ -1,318 +1,365 @@
+        public string IMAGE_URI_2 { get; private set; } = "shoes_2.jpg";
+        public string SEARCH_FILTER { get; private set; } = "style=womens";
 
         // For search tests. Product set must be indexed for search to succeed.
-        private const string INDEXED_PRODUCT_SET = "indexed_product_set_id_for_testing";
-        private const string INDEXED_PRODUCT_1 = "indexed_product_id_for_testing_1";
+        public string INDEXED_PRODUCT_SET { get; private set; } = "indexed_product_set_id_for_testing";
+        public string INDEXED_PRODUCT_1 { get; private set; } = "indexed_product_id_for_testing_1";
+        public readonly string PROJECT_ID = Environment.GetEnvironmentVariable("GOOGLE_PROJECT_ID");
 
-        private readonly string _projectId = Environment.GetEnvironmentVariable("GOOGLE_PROJECT_ID");
-
-        readonly CommandLineRunner _productSearch = new CommandLineRunner()
+        private readonly CommandLineRunner _productSearch = new CommandLineRunner()
         {
             Main = ProductSearchProgram.Main,
             Command = "Product Search"
         };
 
-        public ProductSearchTest()
+        // Keep a list of all the things created while running tests.
+        private readonly List<string[]> _createCommands = new List<string[]>();
+
+        /// <summary>
+        ///  Run the command and track all cloud assets that were created.
+        /// </summary>
+        /// <param name="arguments">The command arguments.</param>
+        public ConsoleOutput Run(params string[] arguments)
         {
-            // Create a indexed product set for TestProductSearch() and TestProductSearchGcs()
-            // tests. These tests remain in the project after the test completes.
-            var output = _productSearch.Run("list_product_sets", _projectId, REGION_NAME);
-            if (!output.Stdout.Contains(INDEXED_PRODUCT_SET))
+            if (arguments[0].StartsWith("create_"))
             {
-                _productSearch.Run("create_product_set", _projectId, REGION_NAME, INDEXED_PRODUCT_SET, PRODUCT_SET_DISPLAY_NAME);
+                _createCommands.Add(arguments);
             }
+            return _productSearch.Run(arguments);
+        }
 
-            output = _productSearch.Run("list_products", _projectId, REGION_NAME);
-            if (!output.Stdout.Contains(INDEXED_PRODUCT_1))
+        /// <summary>
+        /// Delete all the things created in Run() commands.
+        /// </summary>
+        protected void DeleteCreations()
+        {
+            // Clean up everything the test created.
+            _createCommands.Reverse();
+            var exceptions = new List<Exception>();
+            foreach (string[] arguments in _createCommands)
             {
-                _productSearch.Run("create_product", _projectId, REGION_NAME, INDEXED_PRODUCT_1, PRODUCT_DISPLAY_NAME, PRODUCT_CATEGORY);
+                var deleteCommand = arguments.ToList();
```

Comment:
Why ToList() here and then back ToArray() again?

Reply:
Fixed.